File: web/src/entities/DesignCanvas.ts
import { DeviceType, CanvasState, CanvasDimensions, CanvasMetadata, ValidationError } from '../types'
import { Layer } from './Layer'

export interface DesignCanvasData {
  id: string
  deviceType: DeviceType
  dimensions: CanvasDimensions
  layers: any[] // Will be Layer instances or layer data
  metadata: CanvasMetadata
  state: CanvasState
}

export class DesignCanvas {
  public readonly id: string
  public deviceType: DeviceType
  public dimensions: CanvasDimensions
  public layers: Layer[]
  public metadata: CanvasMetadata
  public state: CanvasState

  constructor(data: DesignCanvasData) {
    this.validateCanvasData(data)
    
    this.id = data.id
    this.deviceType = data.deviceType
    this.dimensions = data.dimensions
    this.layers = this.processLayers(data.layers)
    this.metadata = data.metadata
    this.state = data.state
  }

  private validateCanvasData(data: DesignCanvasData): void {
    // Validate Canvas ID
    if (!data.id || typeof data.id !== 'string' || data.id.trim() === '') {
      throw new ValidationError('Canvas ID must be a valid non-empty string', 'id')
    }

    // Validate Device Type
    if (!Object.values(DeviceType).includes(data.deviceType)) {
      throw new ValidationError('Device type must be a supported device specification', 'deviceType')
    }

    // Validate Dimensions
    if (!data.dimensions || 
        typeof data.dimensions.width !== 'number' || data.dimensions.width <= 0 ||
        typeof data.dimensions.height !== 'number' || data.dimensions.height <= 0 ||
        typeof data.dimensions.pixelDensity !== 'number' || data.dimensions.pixelDensity <= 0) {
      throw new ValidationError('Canvas dimensions must have positive width, height, and pixelDensity', 'dimensions')
    }

    // Validate aspect ratio for device type
    this.validateDeviceAspectRatio(data.deviceType, data.dimensions)

    // Validate layers array
    if (!Array.isArray(data.layers)) {
      throw new ValidationError('Layers must be an array', 'layers')
    }

    // Validate metadata
    if (!data.metadata || !data.metadata.createdAt || !data.metadata.modifiedAt) {
      throw new ValidationError('Canvas metadata must include createdAt and modifiedAt dates', 'metadata')
    }

    // Validate state
    if (!Object.values(CanvasState).includes(data.state)) {
      throw new ValidationError(`Invalid canvas state: ${data.state}`, 'state')
    }
  }

  private validateDeviceAspectRatio(deviceType: DeviceType, dimensions: CanvasDimensions): void {
    const expectedRatios: Record<DeviceType, number> = {
      [DeviceType.IPHONE_15_PRO]: 393 / 852, // 19.5:9 aspect ratio
      [DeviceType.IPHONE_15_PLUS]: 428 / 926,
      [DeviceType.IPAD_PRO_11]: 834 / 1194,
      [DeviceType.IPAD_PRO_129]: 1024 / 1366,
      [DeviceType.PIXEL_8_PRO]: 448 / 998,
      [DeviceType.GALAXY_S24_ULTRA]: 440 / 956,
      [DeviceType.MACBOOK_PRO_14]: 1512 / 982,
      [DeviceType.DESKTOP_1920X1080]: 1920 / 1080
    }

    const actualRatio = dimensions.width / dimensions.height
    const expectedRatio = expectedRatios[deviceType]
    const tolerance = 0.01 // 1% tolerance for rounding

    if (Math.abs(actualRatio - expectedRatio) > tolerance) {
      throw new ValidationError(
        'Dimensions must maintain accurate aspect ratios for target device',
        'dimensions'
      )
    }
  }

  private processLayers(layersData: any[]): Layer[] {
    // Validate z-index ordering before processing layers
    this.validateZIndexOrdering(layersData)
    
    // Validate at least one layer requirement
    if (!layersData || layersData.length === 0) {
      throw new ValidationError('Canvas must contain at least one layer to be considered valid', 'layers')
    }
    
    const processedLayers: Layer[] = []
    const layerIds = new Set<string>()

    for (let i = 0; i < layersData.length; i++) {
      const layerData = layersData[i]
      
      // If it's already a Layer instance, use it
      if (layerData instanceof Layer) {
        if (layerIds.has(layerData.id)) {
          throw new ValidationError(`Duplicate layer ID: ${layerData.id}`, 'layers')
        }
        layerIds.add(layerData.id)
        processedLayers.push(layerData)
      } else {
        // Create Layer instance from data
        try {
          const layer = new Layer(layerData)
          if (layerIds.has(layer.id)) {
            throw new ValidationError(`Duplicate layer ID: ${layer.id}`, 'layers')
          }
          layerIds.add(layer.id)
          processedLayers.push(layer)
        } catch (error) {
          throw new ValidationError(
            `Invalid layer data at index ${i}: ${error instanceof Error ? error.message : 'Unknown error'}`,
            'layers'
          )
        }
      }
    }

    return processedLayers
  }

  // Validation methods
  public isValid(): boolean {
    try {
      this.validateState()
      return true
    } catch {
      return false
    }
  }

  private validateState(): void {
    // Check that all layers are within canvas bounds
    for (const layer of this.layers) {
      if (!this.isLayerWithinBounds(layer)) {
        throw new ValidationError(`Layer ${layer.id} is outside canvas bounds`)
      }
    }

    // Check for valid z-index ordering
    this.validateZIndexOrdering()
  }

  private isLayerWithinBounds(layer: Layer): boolean {
    const { transform } = layer
    const layerRight = transform.x + (transform.scaleX * 100) // Assuming base width of 100
    const layerBottom = transform.y + (transform.scaleY * 100) // Assuming base height of 100
    
    return transform.x >= 0 && 
           transform.y >= 0 && 
           layerRight <= this.dimensions.width && 
           layerBottom <= this.dimensions.height
  }

  private validateZIndexOrdering(layersData?: any[]): void {
    const dataToValidate = layersData || this.layers
    
    if (dataToValidate.length === 0) return
    
    // Check if layers have zIndex properties and validate ordering
    for (let i = 0; i < dataToValidate.length; i++) {
      const layer = dataToValidate[i]
      if (layer.zIndex !== undefined) {
        if (typeof layer.zIndex !== 'number' || layer.zIndex < 0) {
          throw new ValidationError('Layers must be ordered with valid z-index values', 'layers')
        }
        
        // Check that layers are in ascending z-index order
        if (i > 0 && dataToValidate[i-1].zIndex !== undefined && layer.zIndex < dataToValidate[i-1].zIndex) {
          throw new ValidationError('Layers must be ordered with valid z-index values', 'layers')
        }
      }
    }
  }

  // Utility methods
  public addLayer(layer: Layer): void {
    if (this.layers.some(l => l.id === layer.id)) {
      throw new ValidationError(`Layer with ID ${layer.id} already exists`)
    }
    
    this.layers.push(layer)
    this.updateModifiedAt()
  }

  public removeLayer(layerId: string): boolean {
    const index = this.layers.findIndex(l => l.id === layerId)
    if (index === -1) return false
    
    this.layers.splice(index, 1)
    this.updateModifiedAt()
    return true
  }

  public getLayer(layerId: string): Layer | undefined {
    return this.layers.find(l => l.id === layerId)
  }

  public reorderLayer(layerId: string, newIndex: number): boolean {
    const currentIndex = this.layers.findIndex(l => l.id === layerId)
    if (currentIndex === -1 || newIndex < 0 || newIndex >= this.layers.length) {
      return false
    }
    
    const layer = this.layers.splice(currentIndex, 1)[0]
    this.layers.splice(newIndex, 0, layer)
    this.updateModifiedAt()
    return true
  }

  public updateState(newState: CanvasState): void {
    if (!Object.values(CanvasState).includes(newState)) {
      throw new ValidationError(`Invalid canvas state: ${newState}`)
    }
    this.state = newState
    this.updateModifiedAt()
  }

  private updateModifiedAt(): void {
    this.metadata.modifiedAt = new Date()
  }

  // Serialization
  public toJSON(): Record<string, any> {
    return {
      id: this.id,
      deviceType: this.deviceType,
      dimensions: this.dimensions,
      layers: this.layers.map(layer => layer.toJSON()),
      metadata: {
        ...this.metadata,
        createdAt: this.metadata.createdAt.toISOString(),
        modifiedAt: this.metadata.modifiedAt.toISOString()
      },
      state: this.state
    }
  }

  public static fromJSON(data: any): DesignCanvas {
    return new DesignCanvas({
      ...data,
      metadata: {
        ...data.metadata,
        createdAt: new Date(data.metadata.createdAt),
        modifiedAt: new Date(data.metadata.modifiedAt)
      },
      layers: data.layers // Layer.fromJSON will be called during processing
    })
  }

  // Device simulation utilities
  public getDeviceSpec(): { name: string; category: string; os: string } {
    const deviceSpecs: Record<DeviceType, { name: string; category: string; os: string }> = {
      [DeviceType.IPHONE_15_PRO]: { name: 'iPhone 15 Pro', category: 'phone', os: 'ios' },
      [DeviceType.IPHONE_15_PLUS]: { name: 'iPhone 15 Plus', category: 'phone', os: 'ios' },
      [DeviceType.IPAD_PRO_11]: { name: 'iPad Pro 11"', category: 'tablet', os: 'ios' },
      [DeviceType.IPAD_PRO_129]: { name: 'iPad Pro 12.9"', category: 'tablet', os: 'ios' },
      [DeviceType.PIXEL_8_PRO]: { name: 'Pixel 8 Pro', category: 'phone', os: 'android' },
      [DeviceType.GALAXY_S24_ULTRA]: { name: 'Galaxy S24 Ultra', category: 'phone', os: 'android' },
      [DeviceType.MACBOOK_PRO_14]: { name: 'MacBook Pro 14"', category: 'desktop', os: 'web' },
      [DeviceType.DESKTOP_1920X1080]: { name: 'Desktop 1920x1080', category: 'desktop', os: 'web' }
    }
    
    return deviceSpecs[this.deviceType]
  }

  public setState(newState: CanvasState): void {
    (this.state as any) = newState // Cast needed since state is readonly
    this.updateModifiedAt()
  }

  public clone(): DesignCanvas {
    return DesignCanvas.fromJSON(this.toJSON())
  }
}

File: ios/VYB/Models/DesignCanvas.swift
import Foundation
import CoreData
import SwiftUI

// MARK: - Device Type Enum
enum DeviceType: String, CaseIterable, Codable {
    case iPhone15Pro = "iPhone 15 Pro"
    case iPhone15Plus = "iPhone 15 Plus"
    case iPadPro11 = "iPad Pro 11\""
    case iPadPro129 = "iPad Pro 12.9\""
    case pixel8Pro = "Pixel 8 Pro"
    case galaxyS24Ultra = "Galaxy S24 Ultra"
    case macBookPro14 = "MacBook Pro 14\""
    case desktop1920x1080 = "Desktop 1920x1080"
}

// MARK: - Canvas State Enum
enum CanvasState: String, CaseIterable, Codable {
    case editing = "editing"
    case aiProcessing = "ai-processing"
    case viewing = "viewing"
    case loading = "loading"
}

// MARK: - Canvas Dimensions
struct CanvasDimensions: Codable {
    let width: Double
    let height: Double
    let pixelDensity: Double
}

// MARK: - Canvas Metadata
struct CanvasMetadata: Codable {
    var createdAt: Date
    var modifiedAt: Date
    var tags: [String]
    var description: String?
    var author: String?
}

// MARK: - Core Data Entity
@objc(DesignCanvas)
public class DesignCanvas: NSManagedObject {
    
    // MARK: - Core Data Properties
    @NSManaged public var id: String
    @NSManaged public var deviceTypeRaw: String
    @NSManaged public var dimensionsData: Data
    @NSManaged public var metadataData: Data
    @NSManaged public var stateRaw: String
    @NSManaged public var layers: NSSet?
    
    // MARK: - Computed Properties
    public var deviceType: DeviceType {
        get {
            return DeviceType(rawValue: deviceTypeRaw) ?? .iPhone15Pro
        }
        set {
            deviceTypeRaw = newValue.rawValue
        }
    }
    
    public var dimensions: CanvasDimensions {
        get {
            do {
                return try JSONDecoder().decode(CanvasDimensions.self, from: dimensionsData)
            } catch {
                return CanvasDimensions(width: 393, height: 852, pixelDensity: 3) // Default iPhone 15 Pro
            }
        }
        set {
            do {
                dimensionsData = try JSONEncoder().encode(newValue)
            } catch {
                print("Failed to encode dimensions: \(error)")
            }
        }
    }
    
    public var metadata: CanvasMetadata {
        get {
            do {
                return try JSONDecoder().decode(CanvasMetadata.self, from: metadataData)
            } catch {
                return CanvasMetadata(createdAt: Date(), modifiedAt: Date(), tags: [])
            }
        }
        set {
            do {
                metadataData = try JSONEncoder().encode(newValue)
            } catch {
                print("Failed to encode metadata: \(error)")
            }
        }
    }
    
    public var state: CanvasState {
        get {
            return CanvasState(rawValue: stateRaw) ?? .editing
        }
        set {
            stateRaw = newValue.rawValue
        }
    }
    
    public var layersArray: [Layer] {
        let layersSet = layers as? Set<Layer> ?? []
        return layersSet.sorted { $0.zIndex < $1.zIndex }
    }
}

// MARK: - Validation Methods
extension DesignCanvas {
    
    public func validateCanvasData() throws {
        // Validate Canvas ID
        guard !id.isEmpty else {
            throw ValidationError.invalidCanvasID("Canvas ID must be a valid non-empty string")
        }
        
        // Validate Device Type
        guard DeviceType(rawValue: deviceTypeRaw) != nil else {
            throw ValidationError.invalidDeviceType("Device type must be a supported device specification")
        }
        
        // Validate Dimensions
        guard dimensions.width > 0 && dimensions.height > 0 && dimensions.pixelDensity > 0 else {
            throw ValidationError.invalidDimensions("Canvas dimensions must have positive width, height, and pixelDensity")
        }
        
        // Validate aspect ratio against device specifications
        try validateDeviceAspectRatio()
        
        // Validate layers
        guard let layersSet = layers, layersSet.count > 0 else {
            throw ValidationError.noLayers("Canvas must contain at least one layer to be considered valid")
        }
        
        // Validate z-index ordering
        try validateZIndexOrdering()
    }
    
    private func validateDeviceAspectRatio() throws {
        let expectedRatios: [DeviceType: Double] = [
            .iPhone15Pro: 393.0 / 852.0,
            .iPhone15Plus: 428.0 / 926.0,
            .iPadPro11: 834.0 / 1194.0,
            .iPadPro129: 1024.0 / 1366.0,
            .pixel8Pro: 448.0 / 998.0,
            .galaxyS24Ultra: 440.0 / 956.0,
            .macBookPro14: 1512.0 / 982.0,
            .desktop1920x1080: 1920.0 / 1080.0
        ]
        
        let actualRatio = dimensions.width / dimensions.height
        let expectedRatio = expectedRatios[deviceType] ?? (393.0 / 852.0)
        let tolerance = 0.01
        
        if abs(actualRatio - expectedRatio) > tolerance {
            throw ValidationError.aspectRatioMismatch("Dimensions must maintain accurate aspect ratios for target device")
        }
    }
    
    private func validateZIndexOrdering() throws {
        let layersArray = self.layersArray
        
        for i in 1..<layersArray.count {
            if layersArray[i].zIndex < layersArray[i-1].zIndex {
                throw ValidationError.invalidZIndex("Layers must be ordered with valid z-index values")
            }
        }
    }
}

// MARK: - Convenience Methods
extension DesignCanvas {
    
    @discardableResult
    public func setState(_ newState: CanvasState) -> DesignCanvas {
        state = newState
        var currentMetadata = metadata
        currentMetadata.modifiedAt = Date()
        metadata = currentMetadata
        return self
    }
    
    public func addLayer(_ layer: Layer) throws {
        // Validate layer doesn't already exist
        if layersArray.contains(where: { $0.id == layer.id }) {
            throw ValidationError.duplicateLayer("Layer with ID \(layer.id) already exists")
        }
        
        // Set appropriate z-index
        let maxZIndex = layersArray.map { $0.zIndex }.max() ?? -1
        layer.zIndex = Int32(maxZIndex + 1)
        layer.canvas = self
        
        // Update modified date
        var currentMetadata = metadata
        currentMetadata.modifiedAt = Date()
        metadata = currentMetadata
    }
    
    public func removeLayer(withId layerId: String) throws {
        guard let layer = layersArray.first(where: { $0.id == layerId }) else {
            throw ValidationError.layerNotFound("Layer with ID \(layerId) not found")
        }
        
        // Ensure we don't remove the last layer
        if layersArray.count <= 1 {
            throw ValidationError.noLayers("Canvas must contain at least one layer to be considered valid")
        }
        
        managedObjectContext?.delete(layer)
        
        // Update modified date
        var currentMetadata = metadata
        currentMetadata.modifiedAt = Date()
        metadata = currentMetadata
    }
    
    public func getLayer(withId layerId: String) -> Layer? {
        return layersArray.first { $0.id == layerId }
    }
}

// MARK: - Device Specifications
extension DesignCanvas {
    
    public struct DeviceSpec {
        let name: String
        let category: String
        let os: String
    }
    
    public func getDeviceSpec() -> DeviceSpec {
        let deviceSpecs: [DeviceType: DeviceSpec] = [
            .iPhone15Pro: DeviceSpec(name: "iPhone 15 Pro", category: "phone", os: "ios"),
            .iPhone15Plus: DeviceSpec(name: "iPhone 15 Plus", category: "phone", os: "ios"),
            .iPadPro11: DeviceSpec(name: "iPad Pro 11\"", category: "tablet", os: "ios"),
            .iPadPro129: DeviceSpec(name: "iPad Pro 12.9\"", category: "tablet", os: "ios"),
            .pixel8Pro: DeviceSpec(name: "Pixel 8 Pro", category: "phone", os: "android"),
            .galaxyS24Ultra: DeviceSpec(name: "Galaxy S24 Ultra", category: "phone", os: "android"),
            .macBookPro14: DeviceSpec(name: "MacBook Pro 14\"", category: "desktop", os: "web"),
            .desktop1920x1080: DeviceSpec(name: "Desktop 1920x1080", category: "desktop", os: "web")
        ]
        
        return deviceSpecs[deviceType] ?? DeviceSpec(name: "iPhone 15 Pro", category: "phone", os: "ios")
    }
}

// MARK: - JSON Serialization
extension DesignCanvas {
    
    public func toJSON() -> [String: Any] {
        return [
            "id": id,
            "deviceType": deviceType.rawValue,
            "dimensions": [
                "width": dimensions.width,
                "height": dimensions.height,
                "pixelDensity": dimensions.pixelDensity
            ],
            "layers": layersArray.map { $0.toJSON() },
            "metadata": [
                "createdAt": ISO8601DateFormatter().string(from: metadata.createdAt),
                "modifiedAt": ISO8601DateFormatter().string(from: metadata.modifiedAt),
                "tags": metadata.tags,
                "description": metadata.description as Any,
                "author": metadata.author as Any
            ],
            "state": state.rawValue
        ]
    }
    
    public static func fromJSON(_ json: [String: Any], in context: NSManagedObjectContext) throws -> DesignCanvas {
        guard let id = json["id"] as? String,
              let deviceTypeString = json["deviceType"] as? String,
              let deviceType = DeviceType(rawValue: deviceTypeString),
              let dimensionsDict = json["dimensions"] as? [String: Any],
              let width = dimensionsDict["width"] as? Double,
              let height = dimensionsDict["height"] as? Double,
              let pixelDensity = dimensionsDict["pixelDensity"] as? Double,
              let stateString = json["state"] as? String,
              let state = CanvasState(rawValue: stateString),
              let metadataDict = json["metadata"] as? [String: Any],
              let createdAtString = metadataDict["createdAt"] as? String,
              let modifiedAtString = metadataDict["modifiedAt"] as? String,
              let tags = metadataDict["tags"] as? [String] else {
            throw ValidationError.invalidJSONData("Invalid JSON data for DesignCanvas")
        }
        
        let canvas = DesignCanvas(context: context)
        canvas.id = id
        canvas.deviceType = deviceType
        canvas.dimensions = CanvasDimensions(width: width, height: height, pixelDensity: pixelDensity)
        canvas.state = state
        
        let formatter = ISO8601DateFormatter()
        let createdAt = formatter.date(from: createdAtString) ?? Date()
        let modifiedAt = formatter.date(from: modifiedAtString) ?? Date()
        
        canvas.metadata = CanvasMetadata(
            createdAt: createdAt,
            modifiedAt: modifiedAt,
            tags: tags,
            description: metadataDict["description"] as? String,
            author: metadataDict["author"] as? String
        )
        
        return canvas
    }
}

// MARK: - Validation Errors
enum ValidationError: LocalizedError {
    case invalidCanvasID(String)
    case invalidDeviceType(String)
    case invalidDimensions(String)
    case aspectRatioMismatch(String)
    case invalidZIndex(String)
    case noLayers(String)
    case duplicateLayer(String)
    case layerNotFound(String)
    case invalidJSONData(String)
    
    var errorDescription: String? {
        switch self {
        case .invalidCanvasID(let message),
             .invalidDeviceType(let message),
             .invalidDimensions(let message),
             .aspectRatioMismatch(let message),
             .invalidZIndex(let message),
             .noLayers(let message),
             .duplicateLayer(let message),
             .layerNotFound(let message),
             .invalidJSONData(let message):
            return message
        }
    }
}

// MARK: - Core Data Generated Accessors
extension DesignCanvas {

    @objc(addLayersObject:)
    @NSManaged public func addToLayers(_ value: Layer)

    @objc(removeLayersObject:)
    @NSManaged public func removeFromLayers(_ value: Layer)

    @objc(addLayers:)
    @NSManaged public func addToLayers(_ values: NSSet)

    @objc(removeLayers:)
    @NSManaged public func removeFromLayers(_ values: NSSet)
}